src/fairness/weighted_tree/weighted_tree.cpp

```diff
@@ -109,7 +109,7 @@ void weighted_tree_node_t::propagate_subtree_leaf_size ()
  ******************************************************************************/
 
 weighted_tree_node_t::weighted_tree_node_t (
-    weighted_tree_node_t *parent, const std::string &name,
+    std::shared_ptr<weighted_tree_node_t> parent, const std::string &name,
```

Comment:
@cmoussa1 and @SteVwonder: I vaguely remember there was some memory leak issues with using the shared pointer. I would run valgrind on a test program.

Reply:
Here is some output from running valgrind:

```
[moussa1:test]$ valgrind --leak-check=yes ./mybinary 
==53845== Memcheck, a memory error detector
==53845== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==53845== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==53845== Command: ./mybinary
==53845== 
testing with small_no_tie.db:
leaf.3.1
leaf.3.2
leaf.2.1
leaf.2.2
leaf.1.3
leaf.1.1
leaf.1.2
==53845== 
==53845== HEAP SUMMARY:
==53845==     in use at exit: 2,047 bytes in 26 blocks
==53845==   total heap usage: 306 allocs, 280 frees, 168,237 bytes allocated
==53845== 
==53845== 2,047 (136 direct, 1,911 indirect) bytes in 1 blocks are definitely lost in loss record 8 of 8
==53845==    at 0x4C2A593: operator new(unsigned long) (vg_replace_malloc.c:342)
==53845==    by 0x405A67: __gnu_cxx::new_allocator<std::_Sp_counted_ptr_inplace<Flux::accounting::weighted_tree_node_t, std::allocator<Flux::accounting::weighted_tree_node_t>, (__gnu_cxx::_Lock_policy)2> >::allocate(unsigned long, void const*) (new_allocator.h:104)
==53845==    by 0x405797: std::allocator_traits<std::allocator<std::_Sp_counted_ptr_inplace<Flux::accounting::weighted_tree_node_t, std::allocator<Flux::accounting::weighted_tree_node_t>, (__gnu_cxx::_Lock_policy)2> > >::allocate(std::allocator<std::_Sp_counted_ptr_inplace<Flux::accounting::weighted_tree_node_t, std::allocator<Flux::accounting::weighted_tree_node_t>, (__gnu_cxx::_Lock_policy)2> >&, unsigned long) (alloc_traits.h:357)
==53845==    by 0x40E10D: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<Flux::accounting::weighted_tree_node_t, std::allocator<Flux::accounting::weighted_tree_node_t>, std::shared_ptr<Flux::accounting::weighted_tree_node_t>&, std::string const&, bool, long long, int>(std::_Sp_make_shared_tag, Flux::accounting::weighted_tree_node_t*, std::allocator<Flux::accounting::weighted_tree_node_t> const&, std::shared_ptr<Flux::accounting::weighted_tree_node_t>&, std::string const&, bool&&, long long&&, int&&) (shared_ptr_base.h:616)
==53845==    by 0x40DF4F: std::__shared_ptr<Flux::accounting::weighted_tree_node_t, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<Flux::accounting::weighted_tree_node_t>, std::shared_ptr<Flux::accounting::weighted_tree_node_t>&, std::string const&, bool, long long, int>(std::_Sp_make_shared_tag, std::allocator<Flux::accounting::weighted_tree_node_t> const&, std::shared_ptr<Flux::accounting::weighted_tree_node_t>&, std::string const&, bool&&, long long&&, int&&) (shared_ptr_base.h:1090)
==53845==    by 0x40DDED: std::shared_ptr<Flux::accounting::weighted_tree_node_t>::shared_ptr<std::allocator<Flux::accounting::weighted_tree_node_t>, std::shared_ptr<Flux::accounting::weighted_tree_node_t>&, std::string const&, bool, long long, int>(std::_Sp_make_shared_tag, std::allocator<Flux::accounting::weighted_tree_node_t> const&, std::shared_ptr<Flux::accounting::weighted_tree_node_t>&, std::string const&, bool&&, long long&&, int&&) (shared_ptr.h:316)
==53845==    by 0x40DC77: std::shared_ptr<Flux::accounting::weighted_tree_node_t> std::allocate_shared<Flux::accounting::weighted_tree_node_t, std::allocator<Flux::accounting::weighted_tree_node_t>, std::shared_ptr<Flux::accounting::weighted_tree_node_t>&, std::string const&, bool, long long, int>(std::allocator<Flux::accounting::weighted_tree_node_t> const&, std::shared_ptr<Flux::accounting::weighted_tree_node_t>&, std::string const&, bool&&, long long&&, int&&) (shared_ptr.h:588)
==53845==    by 0x40D9EF: std::shared_ptr<Flux::accounting::weighted_tree_node_t> std::make_shared<Flux::accounting::weighted_tree_node_t, std::shared_ptr<Flux::accounting::weighted_tree_node_t>&, std::string const&, bool, long long, int>(std::shared_ptr<Flux::accounting::weighted_tree_node_t>&, std::string const&, bool&&, long long&&, int&&) (shared_ptr.h:604)
==53845==    by 0x40C3C9: get_sub_banks(sqlite3*, std::string const&, std::shared_ptr<Flux::accounting::weighted_tree_node_t>, sqlite3_stmt*, sqlite3_stmt*, sqlite3_stmt*) (load_accounting_db.cpp:97)
==53845==    by 0x40D622: load_accounting_db(std::string const&) (load_accounting_db.cpp:382)
==53845==    by 0x402326: main (test_load_accounting_db.cpp:41)
==53845== 
==53845== LEAK SUMMARY:
==53845==    definitely lost: 136 bytes in 1 blocks
==53845==    indirectly lost: 1,911 bytes in 25 blocks
==53845==      possibly lost: 0 bytes in 0 blocks
==53845==    still reachable: 0 bytes in 0 blocks
==53845==         suppressed: 0 bytes in 0 blocks
==53845== 
==53845== For lists of detected and suppressed errors, rerun with: -s
==53845== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

Reply:
> I vaguely remember there was some memory leak issues with using the shared pointer.

Ah! Good call.  When all references to the root tree vertex are gone, we want the whole tree to be destructed, but if the children have shared pointers to their parent, then we have a circular reference (parent -> children and children -> parent). 

> How is this typically handled?

It depends on what is causing the memory leak.  It could be a missed free or bad destructor.  In this case, it looks to be a circular reference.  One way to handle this one is to use a `weak_ptr` for the references that children have to their parent:
```diff
diff --git a/src/fairness/weighted_tree/weighted_tree.cpp b/src/fairness/weighted_tree/weighted_tree.cpp
index a1ab4c7..7e27b0b 100644
--- a/src/fairness/weighted_tree/weighted_tree.cpp
+++ b/src/fairness/weighted_tree/weighted_tree.cpp
@@ -87,18 +87,18 @@ void weighted_tree_node_t::calc_set_children_weight ()

 void weighted_tree_node_t::propagate_subtree_size ()
 {
-    if (m_parent == nullptr)
+    if (m_parent.expired())
         return;
-    m_parent->m_subtree_size++;
-    m_parent->propagate_subtree_size ();
+    m_parent.lock()->m_subtree_size++;
+    m_parent.lock()->propagate_subtree_size ();
 }

 void weighted_tree_node_t::propagate_subtree_leaf_size ()
 {
-    if (m_parent == nullptr)
+    if (m_parent.expired())
         return;
-    m_parent->m_subtree_leaf_size++;
-    m_parent->propagate_subtree_leaf_size ();
+    m_parent.lock()->m_subtree_leaf_size++;
+    m_parent.lock()->propagate_subtree_leaf_size ();
 }


@@ -150,7 +150,7 @@ std::shared_ptr<weighted_tree_node_t> weighted_tree_node_t
 std::shared_ptr<weighted_tree_node_t> weighted_tree_node_t
                                           ::get_parent () const
 {
-    return m_parent;
+    return m_parent.lock();
 }

 bool weighted_tree_node_t::is_tie_with_next () const
@@ -241,11 +241,11 @@ int weighted_tree_node_t::dprint_csv (std::ostringstream &out,
     int rc = 0;
     try {
         if (is_user ()) {
-            if (!m_parent) {
+            if (m_parent.expired()) {
                 errno = EINVAL;
                 return -1;
             }
-            out << level << "," << m_parent->get_name () << ","
+            out << level << "," << m_parent.lock()->get_name () << ","
                 << get_name () << "," << get_shares () << "," << get_usage ();
             if (long_format)
                 out << "," << get_fshare();
diff --git a/src/fairness/weighted_tree/weighted_tree.hpp b/src/fairness/weighted_tree/weighted_tree.hpp
index e4939ad..618ffc6 100644
--- a/src/fairness/weighted_tree/weighted_tree.hpp
+++ b/src/fairness/weighted_tree/weighted_tree.hpp
@@ -65,7 +65,7 @@ private:
     uint64_t m_subtree_leaf_size = 0;
     double m_weight = 0.0f;
     bool m_tie_with_next = false;
-    std::shared_ptr<weighted_tree_node_t> m_parent = nullptr;
+    std::weak_ptr<weighted_tree_node_t> m_parent = std::weak_ptr<weighted_tree_node_t>();
     std::vector<std::shared_ptr<weighted_tree_node_t>> m_children;
 };
```
I **think** that is the best way to handle this, but I'll defer to @dongahn as to what is best.

Reply:
PS - I didn't get a chance to run that through valgrind, so it may not completely solve the memory leak.

Reply:
> It depends on what is causing the memory leak. It could be a missed free or bad destructor. In this case, it looks to be a circular reference. One way to handle this one is to use a weak_ptr for the references that children have to their parent:

I remember using a raw pointer (as before) was the best I could come up with. I do remember I played with destructors and stuff but couldn't find a reasonable combination for this at all.

Reply:
Thanks for explaining this @SteVwonder and @dongahn - I went ahead and made the suggested changes and re-ran valgrind with my test program:

```
[moussa1@quartz386:test]$ valgrind --leak-check=yes ./mybinary 
==2277== Memcheck, a memory error detector
==2277== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==2277== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==2277== Command: ./mybinary
==2277== 
testing with small_no_tie.db:
leaf.3.1
leaf.3.2
leaf.2.1
leaf.2.2
leaf.1.3
leaf.1.1
leaf.1.2
==2277== 
==2277== HEAP SUMMARY:
==2277==     in use at exit: 0 bytes in 0 blocks
==2277==   total heap usage: 306 allocs, 306 frees, 168,237 bytes allocated
==2277== 
==2277== All heap blocks were freed -- no leaks are possible
==2277== 
==2277== For lists of detected and suppressed errors, rerun with: -s
==2277== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

So it looks like the leaks are gone now, so I can push those changes up.